### PR TITLE
Feature/tableau api cd

### PR DESCRIPTION
**Pull request by Caleb-Dixon**
adds front end for management of api keys. Includes seeing existing keys, disabling keys, and adds readable names to keys

### CODE REVIEW

File: frontend/templates/manage_credentials.html [hunk 1]
```diff
@@ -1,90 +1,143 @@
...
+              <form ng-if="api_key.is_active" action="/disable_api_key" method="post">
+                <div style="display:none;" class=form-group>
+                  <input type="text" class="form-control" name="api_key_id" id="field_{% raw %}{{ api_key.access_key_id }}{% endraw %}" value="{% raw %}{{ api_key.access_key_id }}{% endraw %}"/>
+                </div>
+                <button type="submit" class="btn active btn-primary" onclick="return confirm('You are about to permanently disable this API Key for all users. Are you sure you want to do this?')">Disable Key</button>
+              </form>
+              <div ng-if="!api_key.is_active">Disabled</div>
+            </td>
+          </tr>
+        <tbody/>
```
Comment (alvinsiu) [inline frontend/templates/manage_credentials.html:136]: ;]

File: frontend/templates/manage_researchers.html [hunk 2]
```diff
@@ -46,8 +46,7 @@ <h3>All Researchers</h3>
                         </span>
                     </td>
                     <td>
-                        <a href="/edit_researcher/{% raw %}{{ researcher_data[0].id }}{% endraw %}">edit researcher</a>
-                    </td>
+e                    </td>
```
Comment (alvinsiu) [inline frontend/templates/manage_researchers.html:49]: This might be a typo?

File: pages/admin_pages.py [hunk 3]
```diff
@@ -193,3 +199,46 @@ def reset_download_api_credentials():
...
+        <p>Your new <b>Access Key</b> is:
+          <div class="container-fluid">
+            <textarea rows="1" cols="85" readonly="readonly" onclick="this.focus();this.select()">%s</textarea></p>
+          </div>
+        <p>Your new <b>Secret Key</b> is:
+          <div class="container-fluid">
+            <textarea rows="1" cols="85" readonly="readonly" onclick="this.focus();this.select()">%s</textarea></p>
+          </div>
+        <p>Please record these somewhere; This secret key will not be shown again!</p>""" \
+          % (api_key.access_key_id, api_key.access_key_secret_plaintext)
```
Comment (alvinsiu) [inline pages/admin_pages.py:221]: Suggestion:
```
message = f"""
    <h3>...
    ...>{api_key.api_key_id}</textarea>...
    ...>{api_key.api_key_secret_plaintext}</textarea>...
"""
```

File: pages/admin_pages.py [hunk 4]
```diff
@@ -193,3 +199,46 @@ def reset_download_api_credentials():
...
+    api_key.save()
+    flash(Markup(msg), 'warning')
+    return redirect("/manage_credentials")
```
Comment (alvinsiu) [inline pages/admin_pages.py:224]: Can we use `url_for` or something here too?

File: pages/admin_pages.py [hunk 5]
```diff
@@ -193,3 +199,46 @@ def reset_download_api_credentials():
...
+    if not api_key_query.exists():
+        flash(Markup("No matching API key found to disable"), 'warning')
+        return redirect("/manage_credentials")
+    api_key = api_key_query[0]
+    if not api_key.is_active:
+        flash("That API key has already been disabled: %s" % api_key_id, 'warning')
+        return redirect("/manage_credentials")
+    api_key.is_active = False
+    api_key.save()
+    return redirect("/manage_credentials")
```
Comment (alvinsiu) [inline pages/admin_pages.py:244]: Can we formulate this as a `DELETE` request?

File: pages/admin_pages.py [hunk 6]
```diff
@@ -193,3 +199,46 @@ def reset_download_api_credentials():
...
+    api_key_query = ApiKey.objects.filter(access_key_id=api_key_id).filter(researcher=get_session_researcher())
+    if not api_key_query.exists():
+        flash(Markup("No matching API key found to disable"), 'warning')
+        return redirect("/manage_credentials")
+    api_key = api_key_query[0]
+    if not api_key.is_active:
+        flash("That API key has already been disabled: %s" % api_key_id, 'warning')
+        return redirect("/manage_credentials")
+    api_key.is_active = False
+    api_key.save()
```
Comment (alvinsiu) [inline pages/admin_pages.py:243]: You could also collapse this whole thing into:
`ApiKey.objects.filter(researcher=get_session_researcher(), access_key_id=api_key_id, is_active=True).update(is_active=False)`

File: pages/admin_pages.py [hunk 7]
```diff
@@ -193,3 +199,46 @@ def reset_download_api_credentials():
...
+          % (api_key.access_key_id, api_key.access_key_secret_plaintext)
+    api_key.save()
+    flash(Markup(msg), 'warning')
+    return redirect("/manage_credentials")
+
+
+@admin_pages.route('/disable_api_key', methods=['POST'])
+@authenticate_researcher_login
+def disable_api_key():
+    if "api_key_id" not in request.values:
```
Comment (alvinsiu) [inline pages/admin_pages.py:230]: I'd also use a Django form here

File: pages/admin_pages.py (hunk 5 above)
Comment (Caleb-Dixon) [reply pages/admin_pages.py:244]: When researching how to do this, I found out that HTML5 explicitly does not support methods other than GET and POST in forms. This [stack exchange thread] (https://softwareengineering.stackexchange.com/questions/114156/why-are-there-are-no-put-and-delete-methods-on-html-forms) goes into it, but it seems like a contentious topic with a some history (at one point it was implemented, then removed). There are ways to do it, but it seems to me that the design intention for HTML is for deletions to be done with POST requests and would be more inclined to leave it as is. What is your opinion?

Comment (alvinsiu) [reply pages/admin_pages.py:244]: Hmmm, I think this point resonates most with me:
```
HTML is inextricably bound to HTTP. HTML is the human interface of HTTP. It's therefore automatically questionable why HTML does not support all relevant methods in the HTTP specification. Why can machines PUT and DELETE resources, but humans cannot?
```

So my take would be to keep the `DELETE` request on the backend and just switch to using a JS request to handle deletes (i.e. `Executing PUT and DELETE to modify resources on the origin server is straight-forward for modern Web browsers using the XmlHttpRequest object`), though the notable problem with this is if a browser has JS turned off.

However, the other complication is doing it this way means you also need to handle injecting the CSRF token into the JS request, which can range from fairly easy to not very trivial.

As much as I want to keep the semantics of using a `DELETE` request, I think the easier solution would be to keep the endpoint as something like `api-key/delete` and just use `POST`.

**Comment by Caleb-Dixon** [comment]
add csrf token to all forms

**Pull request merged**